src/ndarray/ndarray.cc

```diff
@@ -558,6 +558,101 @@ void CopyFromTo(const NDArray& from, const NDArray& to, int priority) {
   }
 }
 
+void Quantize(const NDArray &from, NDArray *to, NDArray *residual, const std::string& compress,
+              const float neg_threshold, const float pos_threshold,
+              int priority) {
+  CHECK(from.shape().ndim() != 0)
+      << "source operands have zero dimension shape";
+  // important: callback must always capture by value
+  NDArray ret = *to;
+  NDArray res = *residual;
+  int a = from.ctx().dev_mask();
+  int b = to->ctx().dev_mask();
+  if (a == cpu::kDevMask && b == cpu::kDevMask) {
+    if (compress == "2bit") {
+      Engine::Get()->PushSync([from, res, ret, neg_threshold, pos_threshold](RunContext ctx) {
+          std::vector<TBlob> inputs(3);
+          inputs[0] = from.data();
+          inputs[1] = res.data();
+          inputs[2] = ret.data();
+          mxnet::ndarray::Quantize2BitDispatch<cpu>(ctx.get_stream<cpu>(), inputs,
+                                                    neg_threshold, pos_threshold);
+        }, from.ctx(), {from.var()}, {ret.var(), res.var()},
+        FnProperty::kNormal, priority, PROFILER_MESSAGE("QuantizeCPU"));
+    } else {
+      LOG(FATAL) << "Unsupported Quantization";
+    }
+  } else {
+#if MXNET_USE_CUDA
+    if (a == gpu::kDevMask && b == gpu::kDevMask) {
```

Comment:
Can this be combined with the above else into else if?

Reply:
Fair point, will do that